A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,1095 @@
+is necessary because the RPCs are queued by the channel, and the LB policy
+cannot only re-pick some of them.)
+
+The connectivity state of the policy will be the aggregated state of all the
+child policies. Even though this is not true for a specific RPC (the actual
+state is the state of the child policy that will be picked), it represents
+the overall state of the channel.
+
+## xDS resolver and xDS client
+
+The xDS resolver’s overall behavior will be similar as before. It will call
+the xDS client, and will send a service config generated from
+RouteConfiguration (the RDS response) to the parent gRPC client channel.
+
+The xDS client will make an LDS request, with resource\_names set to the
+target URI with “xds:” scheme removed. The client will get
+HttpConnectionManager configuration from the listener response, and make an
+RDS request to get RouteConfiguration.
+
+All new features should be flag-protected until interop testing. Balancers
+are passively picked by service config, so the flag-protection can be done
+before service config generation. Because xds\_client also does response
+validation, and xds\_resolver behavior can be controlled by the parsed result
+from xds\_client, the flag-protection should be done in the xds\_client. The
+flag-protection will use the following environment variable:
+
+```
+GRPC_XDS_EXPERIMENTAL_ROUTING = (true|false)
+```
+
+### Response validation
+
+After LDS/RDS, when the xDS client gets a
+[RouteConfiguration](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route.proto#L24),
+it will look at the list of VirtualHosts and find the element whose
+[domains](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L72)
+field **_best matches_** the server name from URI (with scheme stripped off).
+If no matching VirtualHost is found, the xDS client will report that to the
+xDS resolver, and the xDS resolver will return an error to the client
+channel.
+
+In the matched VirtualHost, the xDS client will iterate over the list of
+[routes](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L76),
+and validate match field and action field:
+
+The route’s [match
+field](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L179)
+
+*   Must have
+    [path\_specifier](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L360)
+    *   Options supported:
+        [prefix](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L365),
+        [path](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L369)
+        or
+        [safe\_regex](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L405)
+    *   NOT regex
+        *   Justification: unsafe and deprecated -- replaced by safe\_regex
```

Comment:
Does Envoy also reject (NACK) regex? If not then we will have a problem when migrating from proxyful to proxyless.